Framework/Algorithms/src/AverageSpectrumBackground.cpp

```diff
@@ -0,0 +1,151 @@
+      Kernel::make_unique<API::WorkspaceProperty<API::MatrixWorkspace>>(
+          "InputWorkspace", "", Kernel::Direction::Input,
+          boost::make_shared<API::CommonBinsValidator>()),
+      "An input workspace.");
+
+  // bottom background range
+  declareProperty(Kernel::make_unique<Kernel::ArrayProperty<size_t>>(
+                      "BottomBackgroundRange", std::vector<size_t>()),
+                  "A list of the bottom background ranges.");
+  // top background range
+  declareProperty(Kernel::make_unique<Kernel::ArrayProperty<size_t>>(
+                      "TopBackgroundRange", std::vector<size_t>()),
+                  "A list of the top background ranges.");
+
+  // Output workspace
+  declareProperty(Kernel::make_unique<API::WorkspaceProperty<>>(
+                      "OutputWorkspace", "", Kernel::Direction::Output),
+                  "A Workspace with the background removed.");
+}
+
+//----------------------------------------------------------------------------------------------
+/** Execute the algorithm.
+ */
+void AverageSpectrumBackground::exec() {
+  API::MatrixWorkspace_sptr inputWS = getProperty("InputWorkspace");
+  std::vector<size_t> bottomBgdRange = getProperty("BottomBackgroundRange");
+  std::vector<size_t> topBgdRange = getProperty("TopBackgroundRange");
+
+  if (bottomBgdRange.empty() && topBgdRange.empty())
+    throw std::invalid_argument("At least one background range is required");
+  if (bottomBgdRange.size() != 2 && !bottomBgdRange.empty())
+    throw std::invalid_argument("BottomBackgroundRange must have length 2");
+  if (topBgdRange.size() != 2 && !topBgdRange.empty())
+    throw std::invalid_argument("TopBackgroundRange must have length 2");
+  if ((!bottomBgdRange.empty() && bottomBgdRange[1] - bottomBgdRange[0] == 0) ||
+      (!topBgdRange.empty() && topBgdRange[1] - topBgdRange[0] == 0)) {
+    throw std::invalid_argument("Cannot have a range of length 0");
+  }
+
+  API::MatrixWorkspace_sptr bottomBgd, topBgd;
+  // groups the bottom background into a single spectra workspace
+  if (!bottomBgdRange.empty()) {
+    bottomBgd =
+        groupBackgroundDetectors(inputWS, getSpectraFromRange(bottomBgdRange));
+  }
+
+  // groups the top background into a single spectra workspace
+  if (!topBgdRange.empty()) {
+    topBgd =
+        groupBackgroundDetectors(inputWS, getSpectraFromRange(topBgdRange));
+  }
+
+  size_t totalBkgRange;
+  API::MatrixWorkspace_sptr bgd;
+  if (topBgdRange.empty()) {
+    bgd = bottomBgd;
+    totalBkgRange = bottomBgdRange[1] - bottomBgdRange[0];
+  } else if (bottomBgdRange.empty()) {
+    bgd = topBgd;
+    totalBkgRange = topBgdRange[1] - topBgdRange[0];
+  } else {
+    bgd = plus(bottomBgd, topBgd);
+    totalBkgRange = (bottomBgdRange[1] - bottomBgdRange[0]) +
+                    (topBgdRange[1] - topBgdRange[0]);
+  }
+
+  // find the average of the background
+  API::MatrixWorkspace_sptr averageBgd =
+      divide(bgd, static_cast<double>(totalBkgRange));
```

Comment:
Could we skip this if we get GroupDetectors to return the average rather than the sum?